File: quickstart_guides/linked_lists/java/ReorderList.java
/*
 * Title: Reorder List
 *
 * Problem:
 *    Given a singly linked list L: L0→L1→…→Ln-1→Ln,
 *   reorder it to: L0→Ln→L1→Ln-1→L2→Ln-2→…
 *
 *   You may not modify the values in the list's nodes, only nodes itself may be
 *   changed.
 *
 * Execution: javac ReorderList.java && java ReorderList
 */



public class ReorderList {

    // Private node class
    private static class Node {
        private int value;
        private Node next;

        private Node(int value) {
            this.value = value;
        }
    }
   public static void reorderList(Node head) {
        if (head != null && head.next != null) {
             step(head, head);
        }
    }

    Node step(Node l, Node r) {
        if (r.next == null) { // reached right most position
            if (l.next != r) { // list has more than two elements
                Node temp = l.next;
                l.next = r;
                r.next = temp;
                return temp;
            }
            return null;
        } else {
            Node newL = step(l, r.next);

            if (newL == null) { // we're already done
                return null;
            }
            if (newL == r) { // odd number of nodes, meet in the middle
                newL.next = null; // terminate the list
                return null;
            }
            else if (newL.next == r) { // even number of nodes, meet back to back
                r.next = null; // terminate the list
                return null;
            } else {    // update middle of the string
                Node temp = newL.next;
                newL.next = r;
                r.next = temp;
                return temp;
            }
        }
    }

    public static void main(String[] args) {
        Node n = new Node(1);
        n.next = new Node(2);
        n.next.next = new Node(3);
        n.next.next.next = new Node(4);
        n.next.next.next.next = new Node(5);

    }
}

File: quickstart_guides/linked_lists/python/reorder_list.py
"""
Title: Reorder list.

Problem:
    Given a singly linked list L: L0→L1→…→Ln-1→Ln, reorder it to:
    L0→Ln→L1→Ln-1→L2→Ln-2→….

    You may not modify the values in the list's nodes, only nodes itself may be
    changed.

Execution: python reorder_list.py
"""
import unittest


class ListNode:
    """Basic node class for linked list."""

    def __init__(self, x):
        self.val = x
        self.next = None


def reorder_list(head: ListNode) -> None:
    """Reorder linked list."""
    if head is None:
        return None

    fast = slow = head
    # Slow will be the middle of the list
    while fast and fast.next:
        fast = fast.next.next
        slow = slow.next

    # Reverse list elements starting from the middle.
    pr, cur = slow, slow.next
    while cur:
        tmp = cur.next
        cur.next, pr = pr, cur
        cur = tmp

    tail = pr
    # Walk from head and tail towards the middle of the list and merge
    # {0: n}, {1: (n-1)} ... { k: (n-k)}
    while tail != slow:
        tmp = tail.next
        tail.next = head.next
        head.next = tail
        head = tail.next
        tail = tmp

    slow.next = None


class TestReverseList(unittest.TestCase):
    """Unit test for reverse list."""

    def test_1(self):
        """Test for 1->2->3->4->5."""
        input_1 = ListNode(1)
        input_1.next = ListNode(2)
        input_1.next.next = ListNode(3)
        input_1.next.next.next = ListNode(4)
        input_1.next.next.next.next = ListNode(5)

        reorder_list(input_1)


if __name__ == "__main__":
    unittest.main()
